ts/sagas/messages.ts

```diff
@@ -0,0 +1,126 @@
+import { ServicePublic } from "../../definitions/backend/ServicePublic";
+import { BackendClient } from "../api/backend";
+import { apiUrlPrefix } from "../config";
+import { MESSAGES_LOAD_REQUEST } from "../store/actions/constants";
+import {
+  loadMessagesFailure,
+  loadMessagesSuccess
+} from "../store/actions/messages";
+import { sessionTokenSelector } from "../store/reducers/session";
+
+export interface MessagesListObject {
+  [key: string]: MessageWithoutContent;
+}
+
+export type MessagesIdsArray = ReadonlyArray<string>;
+
+export interface NormalizedMessagesResponse {
+  messages: MessagesListObject;
+  ids: MessagesIdsArray;
+}
+
+export const INITIAL_NORMALIZED_MESSAGES_RESPONSE: NormalizedMessagesResponse = {
+  messages: {},
+  ids: []
+};
+
+function createNormalizedMessagesResponse(
+  messages: ReadonlyArray<MessageWithoutContent>
+): NormalizedMessagesResponse {
+  return messages.reduce((result, value) => {
+    return {
+      messages: {
+        ...result.messages,
+        [value.id]: {
+          ...value
+        }
+      },
+      ids: [...result.ids, value.id]
+    };
+  }, INITIAL_NORMALIZED_MESSAGES_RESPONSE);
+}
+
+function* loadMessage(sessionToken: string, id: string): Iterator<Effect> {
+  const backendClient = BackendClient(apiUrlPrefix, sessionToken);
+
+  const response:
+    | BasicResponseType<MessageWithContent>
+    | undefined = yield call(backendClient.getMessage, { id });
+
+  if (!response || response.status !== 200) {
+    return response ? response.value : Error();
+  } else {
+    return response.value;
+  }
+}
+
+function* loadService(sessionToken: string, id: string): Iterator<Effect> {
+  const backendClient = BackendClient(apiUrlPrefix, sessionToken);
+
+  const response: BasicResponseType<ServicePublic> | undefined = yield call(
+    backendClient.getService,
+    { id }
+  );
+
+  if (!response || response.status !== 200) {
+    return response ? response.value : Error();
+  } else {
+    return response.value;
+  }
+}
+
+function* loadMessages(): Iterator<Effect> {
+  // Get the token from the state
+  const sessionToken: string | undefined = yield select(sessionTokenSelector);
+
+  if (sessionToken) {
+    const backendClient = BackendClient(apiUrlPrefix, sessionToken);
+
+    const response: BasicResponseType<Messages> | undefined = yield call(
+      backendClient.getMessages,
+      {}
+    );
+
+    /**
+     * If the response is undefined (can't be decoded) or the status is not 200 dispatch a failure action
+     */
+    if (!response || response.status !== 200) {
+      const error: Error = response ? response.value : Error();
+
+      // Dispatch failure action
+      yield put(loadMessagesFailure(error));
+    } else {
+      // Get the list of the messages
+      const normalizedMessagesResponse = createNormalizedMessagesResponse(
+        response.value.items
+      );
+
+      for (const id of normalizedMessagesResponse.ids) {
```

Comment:
this logic is not correct:

1. you should check whether the content of the message has already been downloaded in a previous check before fetching it
2. you should fetch a Service only once (and also check whether you already have cached the service metadata)

Reply:
I am already adding the fetch only if needed logic